jq255.md

```diff
@@ -0,0 +1,1425 @@
+
+**Warning:** Two important caveats apply:
+
+  - The key exchange described here does not, by itself, ensure
+    authentication. An active attacker in position to intercept messages
+    exchanged between the two parties may replace public keys on the
+    wire with values generated by the attacker, and establish distinct
+    secret keys, but both known to the attacker, with each party. The
+    attacker may then decrypt and re-encrypt subsequent messages as they
+    flow. A Diffie-Hellman key exchange may provide security against
+    active attackers only if integrated into an outer protocol that
+    ensures proper authentication (e.g. SSL/TLS uses signatures and
+    certificates on top of a Diffie-Hellman key exchange).
+
+  - While private and public keys used here have the same format as
+    private and public keys used for signatures, it is **not
+    recommended** to use the same private/public key pair for both key
+    exchange and signatures. In general, the two activities call for key
+    pairs with distinct, incompatible lifecycle properties (simply put,
+    keys used for encryption should normally be backed up to prevent
+    data loss in case of key loss, while keys used for signatures must
+    not be backed up, since their value relies in their exclusive
+    control by the purported signer). Moreover, possible interactions
+    between the two schemes are not entirely explored.
+
+The process below describes how a party uses its own key pair (`sk` and
+`Q`) along with the *encoded* public key received from the peer
+(`peer_pk`) to compute the shared secret. The function returns two
+values: the computed secret key (presumably shared with the peer), and a
+Boolean status. The status is set to `False` if the bytes sent by the
+peer cannot be decoded as a proper public key. When the peer bytes are
+invalid, a secret key is still computed, and it is not guessable by
+outsiders. The point of producing such a fake secret key is to allow the
+use of the key exchange in some (rather contrived) scenarios in which
+attackers cannot *see*, but can *alter* public keys in transit, and may
+obtain some extra information from knowledge of whether the altered key
+was valid or not. Computing a fake but indistinguishable secret key can
+provide this extra security feature. In most normal uses of
+Diffie-Hellman, that feature is not needed.
+
+```python
+def ECDH(sk, Q, peer_pk):
+    # Decode the peer public key from the received bytes.
+    peer_pk_good = True
+    peer_Q = public_key_decode(peer_pk)
+    if peer_Q is False:
+        peer_Q = G
+        peer_pk_good = False
+
+    # Compute the shared point.
+    P = point_mul(peer_Q, sk)
+    eP = group_encode(P)
+
+    # Get the two encoded public keys, in lexicographic order.
+    pk1 = public_key_encode(Q)
+    pk2 = peer_pk
+    ipk1 = int.from_bytes(pk1, byteorder='big')
+    ipk2 = int.from_bytes(pk2, byteorder='big')
```

Comment:
big???

Reply:
It's really about lexicographic ordering. In Python, this is most easily done by converting the two values to big integers (with big-endian convention) and then comparing them. In a language such as Rust or Go, you would simply use a loop over the bytes to know which of the two values should go first into the hash function.

Reply:
Gotcha